File: electron/src/renderer/utils/course-code-parser.ts
const SubjectCodes: Record<string, string> = {
  ADA: 'Drama',
  AMU: 'Music',
  AVI: 'Visual Arts',
  BAF: 'Financial Accounting',
  BAT: 'Accounting',
  BBI: 'Business',
  CGC: 'Geography of Canada',
  CHC: 'Canadian History',
  CHV: 'Civics',
  CHW: 'World History',
  CIA: 'Analyzing Current Issues',
  CLN: 'Law',
  ENG: 'English',
  ENL: 'English',
  ESL: 'English as a Second Language',
  FSF: 'Core French',
  FIF: 'French Immersion',
  GLC: 'Career Studies',
  GLS: 'Learning Strategies',
  GPP: 'Leadership',
  HFA: 'Food & Nutrition',
  HFN: 'Food & Nutrition',
  HHS: 'Human Services',
  HIF: 'Individual & Family',
  HIP: 'Psychology',
  HRE: 'Religion',
  HSB: 'Social Sciences',
  HSP: 'Anthropology & Sociology',
  ICS: 'Computer Science',
  ICD: 'Computer Science',
  MCR: 'Functions',
  MCT: 'Mathematics',
  MCV: 'Calculus & Vectors',
  MDM: 'Data Management',
  MEL: 'Mathematics for Work',
  MFM: 'Foundations of Mathematics',
  MHF: 'Advanced Functions',
  MPM: 'Principles of Mathematics',
  MTH: 'Mathematics',
  NBE: 'Indigenous Studies',
  OLC: 'Ontario Literacy Course',
  PPL: 'Healthy Active Living',
  PAD: 'Outdoor Activities',
  PAF: 'Personal Fitness',
  PAI: 'Physical Activities',
  PSK: 'Introductory Kinesiology',
  SBI: 'Biology',
  SCH: 'Chemistry',
  SES: 'Earth & Space Science',
  SNC: 'Science',
  SPH: 'Physics',
  SVN: 'Environmental Science',
  TEJ: 'Computer Engineering',
  TDJ: 'Technological Design',
  TIK: 'Computer Technology',
  TGJ: 'Communications Technology',
  TMJ: 'Manufacturing Technology',
  TTJ: 'Transportation Technology',
  TWJ: 'Construction Technology',
};

const Pathways: Record<string, string> = {
  D: 'Academic',
  P: 'Applied',
  O: 'Open',
  U: 'University',
  M: 'University/College',
  C: 'College',
  E: 'Workplace',
  W: 'Destreamed',
  L: 'Locally Developed',
};

export interface ParsedCourse {
  subjectName: string;
  gradeLevel: string;
  pathway: string;
}

export function parseCourseCode(courseCode: string): ParsedCourse {
  if (!courseCode || courseCode.length < 5) {
    return { subjectName: courseCode || '', gradeLevel: '', pathway: '' };
  }

  // ESL special handling
  if (courseCode.startsWith('ESL') && courseCode.length >= 5) {
    const eslLevel = courseCode[3];
    return { subjectName: 'English as a Second Language', gradeLevel: `Level ${eslLevel}`, pathway: 'Open' };
  }

  const subjectCode = courseCode.substring(0, 3);
  let gradeLevel = '';
  if (courseCode.length >= 4 && /\d/.test(courseCode[3])) {
    const gradeNum = parseInt(courseCode[3], 10);
    gradeLevel = `Grade ${gradeNum + 8}`;
  }

  let pathway = '';
  if (courseCode.length >= 5 && Pathways[courseCode[4]]) {
    pathway = Pathways[courseCode[4]];
  }

  const subjectName = SubjectCodes[subjectCode] || subjectCode;
  return { subjectName, gradeLevel, pathway };
}

export function getCourseDisplayText(courseCode: string): string {
  const { subjectName, gradeLevel, pathway } = parseCourseCode(courseCode);

  if (courseCode.startsWith('ESL') && courseCode.length >= 5) {
    const eslLevel = courseCode[3];
    return `ESL \u2022 Level ${eslLevel}`;
  }

  if (!gradeLevel && !pathway) return subjectName;
  if (gradeLevel && pathway) return `${subjectName} \u2022 ${gradeLevel} ${pathway}`;
  return `${subjectName} \u2022 ${gradeLevel}${pathway}`;
}

File: TeachAssistApp/Helpers/CourseCodeParser.cs
using System.Collections.Generic;

namespace TeachAssistApp.Helpers
{
    public static class CourseCodeParser
    {
        private static readonly Dictionary<string, string> SubjectCodes = new()
        {
            {"ADA", "Drama"},
            {"AMU", "Music"},
            {"AVI", "Visual Arts"},
            {"BAF", "Financial Accounting"},
            {"BAT", "Accounting"},
            {"BBI", "Business"},
            {"CGC", "Geography of Canada"},
            {"CHC", "Canadian History"},
            {"CHV", "Civics"},
            {"CHW", "World History"},
            {"CIA", "Analyzing Current Issues"},
            {"CLN", "Law"},
            {"ENG", "English"},
            {"ENL", "English"},
            {"ESL", "English as a Second Language"},
            {"FSF", "Core French"},
            {"FIF", "French Immersion"},
            {"GLC", "Career Studies"},
            {"GLS", "Learning Strategies"},
            {"GPP", "Leadership"},
            {"HFA", "Food & Nutrition"},
            {"HFN", "Food & Nutrition"},
            {"HHS", "Human Services"},
            {"HIF", "Individual & Family"},
            {"HIP", "Psychology"},
            {"HRE", "Religion"},
            {"HSB", "Social Sciences"},
            {"HSP", "Anthropology & Sociology"},
            {"ICS", "Computer Science"},
            {"ICD", "Computer Science"},
            {"MCR", "Functions"},
            {"MCT", "Mathematics"},
            {"MCV", "Calculus & Vectors"},
            {"MDM", "Data Management"},
            {"MEL", "Mathematics for Work"},
            {"MFM", "Foundations of Mathematics"},
            {"MHF", "Advanced Functions"},
            {"MPM", "Principles of Mathematics"},
            {"MTH", "Mathematics"},
            {"NBE", "Indigenous Studies"},
            {"OLC", "Ontario Literacy Course"},
            {"PPL", "Healthy Active Living"},
            {"PAD", "Outdoor Activities"},
            {"PAF", "Personal Fitness"},
            {"PAI", "Physical Activities"},
            {"PSK", "Introductory Kinesiology"},
            {"SBI", "Biology"},
            {"SCH", "Chemistry"},
            {"SES", "Earth & Space Science"},
            {"SNC", "Science"},
            {"SPH", "Physics"},
            {"SVN", "Environmental Science"},
            {"TEJ", "Computer Engineering"},
            {"TDJ", "Technological Design"},
            {"TIK", "Computer Technology"},
            {"TGJ", "Communications Technology"},
            {"TMJ", "Manufacturing Technology"},
            {"TTJ", "Transportation Technology"},
            {"TWJ", "Construction Technology"}
        };

        private static readonly Dictionary<char, string> Pathways = new()
        {
            {'D', "Academic"},
            {'P', "Applied"},
            {'O', "Open"},
            {'U', "University"},
            {'M', "University/College"},
            {'C', "College"},
            {'E', "Workplace"},
            {'W', "Destreamed"},
            {'L', "Locally Developed"}
        };

        public static (string SubjectName, string GradeLevel, string Pathway) Parse(string courseCode)
        {
            if (string.IsNullOrEmpty(courseCode) || courseCode.Length < 5)
                return (courseCode ?? "", "", "");

            // Handle ESL courses specially (ESLAO, ESLBO, ESLCO, ESLDO, ESLEO)
            if (courseCode.StartsWith("ESL") && courseCode.Length >= 5)
            {
                char eslLevel = courseCode[3]; // A, B, C, D, or E
                return ("English as a Second Language", $"Level {eslLevel}", "Open");
            }

            // Extract parts: ABC#X# (e.g., MTH1W1-8)
            string subjectCode = courseCode.Length >= 3 ? courseCode.Substring(0, 3) : courseCode;

            // Grade number is 4th character
            string gradeLevel = "";
            if (courseCode.Length >= 4 && char.IsDigit(courseCode[3]))
            {
                int gradeNum = courseCode[3] - '0';
                gradeLevel = $"Grade {gradeNum + 8}"; // 1=Grade 9, 2=Grade 10, etc.
            }

            // Pathway is 5th character
            string pathway = "";
            if (courseCode.Length >= 5 && Pathways.ContainsKey(courseCode[4]))
            {
                pathway = Pathways[courseCode[4]];
            }

            // Look up subject name
            string subjectName = SubjectCodes.ContainsKey(subjectCode)
                ? SubjectCodes[subjectCode]
                : subjectCode;

            return (subjectName, gradeLevel, pathway);
        }

        public static string GetDisplayText(string courseCode)
        {
            var (subject, grade, pathway) = Parse(courseCode);

            // Special handling for ESL courses - shorter display
            if (courseCode.StartsWith("ESL") && courseCode.Length >= 5)
            {
                char eslLevel = courseCode[3]; // A, B, C, D, or E
                return $"ESL • Level {eslLevel}";
            }

            if (string.IsNullOrEmpty(grade) && string.IsNullOrEmpty(pathway))
                return subject;

            if (!string.IsNullOrEmpty(grade) && !string.IsNullOrEmpty(pathway))
                return $"{subject} • {grade} {pathway}";

            return $"{subject} • {grade}{pathway}";
        }
    }
}
